File: convert-sorted-array-to-binary-search-tree/convert-sorted-array-to-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def BST(self, nums: list[int], l: int, r: int) -> TreeNode:
        '''
        Performs the recursive D&C to create the balanced BST
        '''
        root = None
        if l <= r:
            mid = (l + r) // 2
            root = TreeNode(nums[mid])
            root.left = self.BST(nums, l, mid - 1)
            root.right = self.BST(nums, mid + 1, r)
        return root
    
    def sortedArrayToBST(self, nums: List[int]) -> TreeNode:
        '''
        Takes a sorted array and converts it to a balanced binary tree
        Returns the root of a balanced binary tree
        '''
        return self.BST(nums, 0, len(nums) - 1)

File: convert-sorted-array-to-binary-search-tree/convert-sorted-array-to-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution 
{
    public TreeNode sortedArrayToBST(int[] nums) 
    {
        // No tree can be built
        if (nums.length < 1) return null;
        
        // Build the tree using DFS
        TreeNode root = DFS(nums, 0, nums.length-1);
        
        return root;
    }
    
    private TreeNode DFS(int[] nums, int low, int high)
    {
        // Finished this subtree
        if (low > high) return null;
        
        // Find mid (avoid integer overflow)
        int mid = low + (high-low) / 2;
        
        // Create a subtree root
        TreeNode root = new TreeNode(nums[mid]);
        
        // Build the subtree child nodes
        root.left = DFS(nums, low, mid-1);
        root.right = DFS(nums, mid+1, high);
        
        // Return the subtree
        return root;
    }
}